### PR TITLE
✨Return Secret Metadata with Secret

### PR DESCRIPTION
**Pull request by dericmiller**
## Context and Description of Changes Proposed
This work adds an accessible Metadata struct to the Vaultlib Secret that the authentication plugin will be able to use as a `kid` when creating a JWT.

## Draft Commit Message Body
```
✨Add Metadata struct to Vaultlib Secret

PR #5
close [ch1305]
```

### CODE REVIEW

File: auth.go [hunk 1]
```diff
@@ -8,6 +8,14 @@ import (
 	"github.com/pkg/errors"
 )
 
+// metadata holds additional information about Vault secrets (kv v2)
+type metadata struct {
```
Comment (jambonrose) [inline auth.go]: Given that `metadata` is used on `Secret`, a public struct, we should also make `metadata` a public struct.

File: vault_test.go [hunk 2]
```diff
@@ -2,6 +2,7 @@ package vaultlib
 
 import (
 	"encoding/json"
+	// "fmt" // ***
```
Comment (jambonrose) [inline vault_test.go]: Please remove this comment.

File: vault_test.go [hunk 3]
```diff
@@ -103,13 +105,18 @@ func TestVaultClient_GetSecret(t *testing.T) {
 		t.Run(tt.name, func(t *testing.T) {
 			c := tt.cli
 			res, err := c.GetSecret(tt.path)
+			// fmt.Println("Vault Test: ", tt.name)            // ***
+			// fmt.Println("Metadata: ", res.Metadata.Version) // ***
```
Comment (jambonrose) [inline vault_test.go]: Please remove these comments.

File: vault_test.go [hunk 4]
```diff
@@ -103,13 +105,18 @@ func TestVaultClient_GetSecret(t *testing.T) {
...
 			if (err != nil) != tt.wantErr {
 				t.Errorf("Client.GetSecret() error = %v, wantErr %v", err, tt.wantErr)
 				return
 			}
 			if !reflect.DeepEqual(res.KV, tt.wantKv) || !reflect.DeepEqual(res.JSONSecret, tt.wantJSON) {
 				t.Errorf("Client.GetSecret() = %v, want %v", res.KV, tt.wantKv)
 			}
+			if (tt.wantMetadataVersion == 0 && res.Metadata.Version != 0) || (tt.wantMetadataVersion != 0 && res.Metadata.Version == 0) {
```
Comment (jambonrose) [inline vault_test.go]: These conditions are fragile. I replaced them with:

```go
if tt.wantMetadataVersion != res.Metadata.Version {
```

However, I now see failures in the tests:

```
Testing with Vault version 1.4.1
TestMain: Preparing Vault server
TestMain: Vault Prepared; Running Tests
--- FAIL: TestVaultClient_GetSecret (12.01s)
    --- FAIL: TestVaultClient_GetSecret/kvv2 (0.00s)
        vault_test.go:118: Metadata Version = 9, want 1
    --- FAIL: TestVaultClient_GetSecret/json-secretV2 (0.00s)
        vault_test.go:118: Metadata Version = 9, want 1
FAIL
exit status 1
FAIL	github.com/TrueTickets/vaultlib	17.978s
```

If I re-run the tests, then the Metadata version jumps from 9 to 10. The problem is that the test data is not being cleaned up from Vault.

I'd like @athak 's help on this front. Specifically, the test setup provides a Vault setup via `integration_test.go` and `test-files/initVaultDev.sh` however there is not teardown / cleanup. Does Vault have any commands that we could run to clear the data?

**Review by jambonrose — CHANGES_REQUESTED** [review]
This is a good start. Let's find a way to improve our tests by cleaning up the test setup and let's clear up a few stray comments.

File: vault_test.go (hunk 4 above)
Comment (athak) [reply vault_test.go]: Vault is being run in dev mode so there is no persistence. So just by stopping the container, it should be good to go. I agree that the test should shut it down/remove it after the run but I understand why you might want to keep it around to inspect the data in it. I think that should be optional behavior and the default should be to clean up after the run.
I can take of that in separate PR if you want.

Comment (athak) [reply vault_test.go]: My bad, there is no container, it's just stopping the process. But running it in a container is also an improvement to be made.

Comment (jambonrose) [reply vault_test.go]: Given that the original author has it running as a process, let's keep it as a process and not move to a container.

I think being in a position to opt-out of clean-up makes a lot of sense. However, let's start with the initial feature: we need to clean up the data. What are you proposing achieve this first goal?

Comment (jambonrose) [reply vault_test.go]: Atha and I spoke over the phone, and he will be opening a separate PR that enhances the test script with test cleanup.

**Review by jambonrose — APPROVED** [review]

**Pull request merged**